### ISSUE TITLE
Is there a comparison guide between Ractive / Handlebars / Mustache?

### ISSUE THREAD

**Issue by matthew-dean** [ractivejs/ractive#2145]
Here's my issue: I want to use Ractive.js on the front-end because it's awesome. That's not an issue, the next part is the issue. The backend is PHP. In some cases, I'd like to share templates between Ractive and PHP; e.g. displaying an invoice to the user in the browser and also emailing that same invoice, bound to data, via PHP.

I know Ractive doesn't have a PHP port. However, the documentation states it shares many features from Mustache and Handlebars. My question is: which ones? What is a "safe" template that can be used by, say, both Handlebars.php and Ractive? It would be great to see a table that lists overlapping keywords or features between Ractive and other mustache-y libraries.

(A follow-up question might be: if Ractive supports many features that Handlebars supports, why does it omit certain ones? Why not be mostly a "superset" of Handlebars so that people can easily port to Ractive? It would be like Handlebars with better Virtual DOM, similar to the relationship between Underscore and Lodash, the latter being a more performant variant of the former with 99% feature overlap.)

**Comment by Madgvox** [comment]
The main issue with porting between Ractive and a simple template renderer such as Handlebars is that the philosophy behind the templates are fundamentally different. Theoretically, you _could_ make a template that can function in both Ractive and Handlebars; you're just severely limiting the power of Ractive by doing so.

The most obvious incompatibility is with the concept of "helpers". Handlebars (and similar) shun the concept of logic in templates, instead preferring to leave that to the javascript side. Ractive does not have an implementation of helpers because they'd be totally redundant, as you can put your logic directly inside the templates! 

```html
{{#with { unread: total - read } }}
  <span class="remaining">You have {{ unread }} unread emails remaining.</span>
  {{#if unread > 0 }}
    <div class="unread-notification-icon"></div>
  {{/if}}
{{/with}}
```

```js
var ractive = new Ractive({
  template: '#myTemplate',
  data: {
    total: 25,
    read: 16
  }
  // No helpers necessary, it just works
});
```

That template is certainly not directly compatible with Handlebars. I'm not even sure if it's indirectly compatible.

The reason that there's no comparison table for Ractive and Mustlebars is because while the syntax looks similar on the surface, Ractive templates are not similar to Handstache templates at all.

tl;dr: If you were to make cross-compatible templates, you'd be giving up events, decorators, transitions, expressions, and components at the very least.

**Comment by MartinKolarik** [comment]
Ractive is backward compatible with mustache, with two exceptions:
 - no lambdas - see [this comment](https://github.com/ractivejs/ractive/issues/142#issuecomment-22948119) for an explanation,
 - variables/sections can only be inside text nodes/attributes, e.g., something like `<{{tagName}}>` is not valid. This is a limitation of using virtual DOM instead of returning an HTML string.

As for Handlebars, we support explicit `{{#with}}`, `{{#each}}`, `{{#if}}`, `{{else}}` (no `else if` though), and `{{#unless}}` sections, as well as special references `@index` and `@key` inside `{{#each}}`.

As @Madgvox pointed out, Ractive has many features Mustache doesn't, so while porting a template from Mustache to Ractive should be fairly easy, it isn't possible vice-versa (without serious limitations).

**Comment by Madgvox** [comment]
> (no `else if` though)

I'm not sure if I'm misunderstanding the context but as of 0.7.3 ractive supports `{{elseif}}` blocks.

**Comment by MartinKolarik** [comment]
@Madgvox Thanks for the heads-up! I totally forgot this is now supported.

**Comment by matthew-dean** [comment]
@MartinKolarik Thanks! Very useful! I'll close as both of you have sufficiently answered my question.
